File: backend/fastapi_app/lin-ke/supply_goods.py
import copy
import json
from typing import Any, Dict, List, Tuple
from urllib.parse import urljoin


def clean_string(value: Any) -> str:
    if value is None:
        return ""
    return str(value).strip()


def entity_text(value: Any) -> str:
    if isinstance(value, dict):
        return clean_string(value.get("text"))
    return clean_string(value)


def bd_city_text(payload: Dict[str, Any]) -> str:
    return entity_text(payload.get("bdCity"))


def parse_packages(value: Any) -> Tuple[Dict[str, Any], bool]:
    if isinstance(value, dict):
        return copy.deepcopy(value), False
    if isinstance(value, str):
        try:
            parsed = json.loads(value)
        except json.JSONDecodeError:
            return {}, True
        if isinstance(parsed, dict):
            return parsed, True
        return {}, True
    return {}, False


def encode_packages(packages: Dict[str, Any], was_string: bool) -> Any:
    if was_string:
        return json.dumps(packages, ensure_ascii=False, separators=(",", ":"))
    return packages


def extract_menu_for_optimization(payload: Dict[str, Any]) -> List[Dict[str, Any]]:
    packages, _ = parse_packages(payload.get("packages"))
    view_list = packages.get("viewList") if isinstance(packages, dict) else []
    if not isinstance(view_list, list):
        return []
    groups: List[Dict[str, Any]] = []
    for group_index, group in enumerate(view_list):
        if not isinstance(group, dict):
            continue
        raw_items = group.get("list") if isinstance(group.get("list"), list) else []
        items = []
        for item_index, item in enumerate(raw_items):
            if not isinstance(item, dict):
                continue
            items.append(
                {
                    "index": item_index,
                    "title": clean_string(item.get("title")),
                    "num": clean_string(item.get("num")),
                    "price": clean_string(item.get("price")),
                }
            )
        groups.append(
            {
                "index": group_index,
                "groupName": clean_string(group.get("groupName")),
                "groupSelectNum": clean_string(group.get("groupSelectNum")),
                "items": items,
            }
        )
    return groups


def apply_menu_optimization(
    payload: Dict[str, Any], optimized: Dict[str, Any]
) -> Tuple[Dict[str, Any], List[Dict[str, Any]]]:
    next_payload = copy.deepcopy(payload)
    packages, was_string = parse_packages(next_payload.get("packages"))
    view_list = packages.get("viewList") if isinstance(packages, dict) else []
    if not isinstance(view_list, list):
        return next_payload, []

    changes: List[Dict[str, Any]] = []
    groups = optimized.get("groups") if isinstance(optimized, dict) else []
    if not isinstance(groups, list):
        return next_payload, []

    for group_update in groups:
        if not isinstance(group_update, dict):
            continue
        group_index = group_update.get("index")
        if not isinstance(group_index, int) or group_index < 0 or group_index >= len(view_list):
            continue
        target_group = view_list[group_index]
        if not isinstance(target_group, dict):
            continue
        new_group_name = clean_string(group_update.get("groupName"))
        old_group_name = clean_string(target_group.get("groupName"))
        if new_group_name and new_group_name != old_group_name:
            target_group["groupName"] = new_group_name
            changes.append(
                {
                    "path": f"packages.viewList[{group_index}].groupName",
                    "before": old_group_name,
                    "after": new_group_name,
                }
            )

        item_updates = group_update.get("items") if isinstance(group_update.get("items"), list) else []
        target_items = target_group.get("list") if isinstance(target_group.get("list"), list) else []
        for item_update in item_updates:
            if not isinstance(item_update, dict):
                continue
            item_index = item_update.get("index")
            if not isinstance(item_index, int) or item_index < 0 or item_index >= len(target_items):
                continue
            target_item = target_items[item_index]
            if not isinstance(target_item, dict):
                continue
            new_title = clean_string(item_update.get("title"))
            old_title = clean_string(target_item.get("title"))
            if new_title and new_title != old_title:
                target_item["title"] = new_title
                changes.append(
                    {
                        "path": f"packages.viewList[{group_index}].list[{item_index}].title",
                        "before": old_title,
                        "after": new_title,
                    }
                )

    next_payload["packages"] = encode_packages(packages, was_string)
    return next_payload, changes


def normalize_supply_goods_for_lin_ke(
    payload: Dict[str, Any], lin_ke_mapping: Dict[str, Any], rb_image_base_url: str = ""
) -> Dict[str, Any]:
    groups = normalize_item_groups(payload.get("packages"))
    merchant_name = (
        clean_string(payload.get("hostName"))
        or entity_text(payload.get("rbhost"))
        or entity_text(payload.get("company"))
        or clean_string(payload.get("hostNameInput"))
    )
    product = {
        "source": {
            "type": "rebuild_supply_goods",
            "id": clean_string(payload.get("SupplyGoodsId")) or clean_string(payload.get("goodsId")),
        },
        "title": clean_string(payload.get("goodsName")),
        "salePrice": parse_number(payload.get("price")),
        "originPrice": parse_number(payload.get("originPrice")),
        "category": {
            "id": clean_string(lin_ke_mapping.get("categoryId")),
            "name": clean_string(lin_ke_mapping.get("categoryName")),
        },
        "productType": parse_int(lin_ke_mapping.get("productType")),
        "grouponType": clean_string(payload.get("majorType")),
        "images": normalize_images(payload.get("mainPic"), rb_image_base_url)
        + normalize_images(payload.get("rbimages"), rb_image_base_url),
        "detailImages": normalize_images(payload.get("detailImages"), rb_image_base_url),
        "description": clean_string(payload.get("details")),
        "features": clean_string(payload.get("goodsFeatures")),
        "stockQty": {"totalStock": parse_int(payload.get("signAmount"))},
        "saleTime": {
            "startDate": clean_string(payload.get("saleBegin")),
            "endDate": clean_string(payload.get("saleUntil")),
        },
        "validityPeriod": {"endDate": clean_string(payload.get("validUntil"))},
        "itemGroups": groups,
        "purchaseNotice": {"additionalNotes": clean_string(payload.get("guideline"))},
        "merchant": {"name": merchant_name},
        "hosts": [{"name": merchant_name}] if merchant_name else [],
        "fieldSources": {
            "linKeProductType": clean_string(lin_ke_mapping.get("mealTypeText")),
            "linKeCategory": clean_string(lin_ke_mapping.get("classificationKey")),
        },
        "missingFields": [],
        "warnings": [],
    }
    product["images"] = dedupe_images(product["images"])
    product["detailImages"] = dedupe_images(product["detailImages"])
    product["missingFields"] = collect_missing_fields(product)
    return product


def normalize_item_groups(packages_value: Any) -> List[Dict[str, Any]]:
    try:
        packages, _ = parse_packages(packages_value)
    except (TypeError, json.JSONDecodeError):
        return []
    view_list = packages.get("viewList") if isinstance(packages, dict) else []
    if not isinstance(view_list, list):
        return []
    groups = []
    for group_index, group in enumerate(view_list):
        if not isinstance(group, dict):
            continue
        raw_items = group.get("list") if isinstance(group.get("list"), list) else []
        items = []
        for item_index, item in enumerate(raw_items):
            if not isinstance(item, dict):
                continue
            items.append(
                {
                    "id": clean_string(item.get("id")) or f"{group_index}-{item_index}",
                    "name": clean_string(item.get("title")),
                    "price": parse_number(item.get("price")),
                    "quantity": {"amount": parse_int(item.get("num")) or 1, "unit": "FEN"},
                }
            )
        groups.append(
            {
                "id": clean_string(group.get("groupId")) or str(group_index),
                "name": clean_string(group.get("groupName")),
                "items": items,
                "selectionRule": {
                    "totalCount": parse_int(group.get("groupSelectNum")) or len(items),
                    "optionCount": len(items),
                },
                "canRepeat": False,
            }
        )
    return groups


def normalize_images(value: Any, rb_image_base_url: str = "") -> List[Dict[str, Any]]:
    if value is None:
        return []
    raw_items = value if isinstance(value, list) else [value]
    images = []
    for index, item in enumerate(raw_items):
        url = ""
        uri = ""
        name = ""
        if isinstance(item, str):
            raw = item.strip()
            if raw.startswith(("http://", "https://")):
                url = raw
            elif rb_image_base_url:
                url = urljoin(rb_image_base_url.rstrip("/") + "/", raw.lstrip("/"))
            else:
                uri = raw
            name = raw.rsplit("/", 1)[-1] if raw else ""
        elif isinstance(item, dict):
            url = clean_string(item.get("url"))
            uri = clean_string(item.get("uri"))
            name = clean_string(item.get("name"))
        if url or uri:
            images.append(
                {
                    "url": url,
                    "uri": uri,
                    "name": name,
                    "sortableOnlyId": uri or url or str(index),
                }
            )
    return images


def parse_number(value: Any):
    text = clean_string(value).replace(",", "")
    if not text:
        return None
    try:
        return float(text)
    except ValueError:
        return None


def parse_int(value: Any):
    if isinstance(value, int):
        return value
    number = parse_number(value)
    if number is None:
        return None
    return int(number)


def dedupe_images(images: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    seen = set()
    result = []
    for image in images:
        key = image.get("uri") or image.get("url")
        if not key or key in seen:
            continue
        seen.add(key)
        result.append(image)
    return result


def collect_missing_fields(product: Dict[str, Any]) -> List[str]:
    missing = []
    if not clean_string(product.get("title")):
        missing.append("title")
    if not product.get("salePrice"):
        missing.append("salePrice")
    if not product.get("originPrice"):
        missing.append("originPrice")
    if not product.get("images"):
        missing.append("images")
    if not product.get("itemGroups"):
        missing.append("itemGroups")
    if not clean_string(product.get("merchant", {}).get("name")):
        missing.append("merchant.name")
    if not product.get("category", {}).get("id"):
        missing.append("category.id")
    if not product.get("productType"):
        missing.append("productType")
    return missing

File: backend/src/service/lin-ke/supply-goods.ts
import {
  cleanString,
  deepClone,
  entityText,
  isRecord,
  parseIntValue,
  parseNumber,
  type JsonRecord,
} from "./utils.ts";

interface ParsedPackages {
  packages: JsonRecord;
  wasString: boolean;
  malformed: boolean;
}

export function bdCityText(payload: JsonRecord): string {
  return entityText(payload.bdCity);
}

function parsePackages(value: unknown): ParsedPackages {
  if (isRecord(value)) {
    return { packages: deepClone(value), wasString: false, malformed: false };
  }
  if (typeof value === "string") {
    try {
      const parsed = JSON.parse(value) as unknown;
      return isRecord(parsed)
        ? { packages: parsed, wasString: true, malformed: false }
        : { packages: {}, wasString: true, malformed: true };
    } catch {
      return { packages: {}, wasString: true, malformed: true };
    }
  }
  return { packages: {}, wasString: false, malformed: false };
}

function encodePackages(packages: JsonRecord, wasString: boolean): unknown {
  return wasString ? JSON.stringify(packages) : packages;
}

export function displaySupplyGoodsPackages(value: unknown): JsonRecord {
  return parsePackages(value).packages;
}

export function applyEditablePackages(
  basePackagesValue: unknown,
  editedPackagesValue: unknown,
  outputTemplateValue: unknown = basePackagesValue,
): { packages: unknown; changes: JsonRecord[] } {
  const parsedBase = parsePackages(basePackagesValue);
  const parsedTemplate = parsePackages(outputTemplateValue);
  const editedPackages = isRecord(editedPackagesValue) ? editedPackagesValue : {};
  const nextPackages = deepClone(parsedBase.packages);
  const baseViewList = Array.isArray(parsedBase.packages.viewList) ? parsedBase.packages.viewList : [];
  const nextViewList = Array.isArray(nextPackages.viewList) ? nextPackages.viewList : [];
  const editedViewList = Array.isArray(editedPackages.viewList) ? editedPackages.viewList : [];
  const changes: JsonRecord[] = [];

  for (const [groupIndex, baseGroup] of baseViewList.entries()) {
    const nextGroup = nextViewList[groupIndex];
    const editedGroup = editedViewList[groupIndex];
    if (!isRecord(baseGroup) || !isRecord(nextGroup) || !isRecord(editedGroup)) continue;

    const groupName = cleanString(editedGroup.groupName);
    const oldGroupName = cleanString(baseGroup.groupName);
    if (groupName && groupName !== oldGroupName) {
      nextGroup.groupName = groupName;
      changes.push({
        path: `packages.viewList[${groupIndex}].groupName`,
        before: oldGroupName,
        after: groupName,
      });
    }

    const baseItems = Array.isArray(baseGroup.list) ? baseGroup.list : [];
    const nextItems = Array.isArray(nextGroup.list) ? nextGroup.list : [];
    const editedItems = Array.isArray(editedGroup.list) ? editedGroup.list : [];
    for (const [itemIndex, baseItem] of baseItems.entries()) {
      const nextItem = nextItems[itemIndex];
      const editedItem = editedItems[itemIndex];
      if (!isRecord(baseItem) || !isRecord(nextItem) || !isRecord(editedItem)) continue;
      const title = cleanString(editedItem.title);
      const oldTitle = cleanString(baseItem.title);
      if (title && title !== oldTitle) {
        nextItem.title = title;
        changes.push({
          path: `packages.viewList[${groupIndex}].list[${itemIndex}].title`,
          before: oldTitle,
          after: title,
        });
      }
    }
  }

  return {
    packages: encodePackages(nextPackages, parsedTemplate.wasString),
    changes,
  };
}

export function extractMenuForOptimization(payload: JsonRecord): JsonRecord[] {
  const { packages } = parsePackages(payload.packages);
  const viewList = Array.isArray(packages.viewList) ? packages.viewList : [];
  const groups: JsonRecord[] = [];

  for (const [groupIndex, group] of viewList.entries()) {
    if (!isRecord(group)) continue;
    const rawItems = Array.isArray(group.list) ? group.list : [];
    const items: JsonRecord[] = [];
    for (const [itemIndex, item] of rawItems.entries()) {
      if (!isRecord(item)) continue;
      items.push({
        index: itemIndex,
        title: cleanString(item.title),
        num: cleanString(item.num),
        price: cleanString(item.price),
      });
    }
    groups.push({
      index: groupIndex,
      groupName: cleanString(group.groupName),
      groupSelectNum: cleanString(group.groupSelectNum),
      items,
    });
  }

  return groups;
}

export function applyMenuOptimization(
  payload: JsonRecord,
  optimized: JsonRecord,
): { payload: JsonRecord; changes: JsonRecord[] } {
  const nextPayload = deepClone(payload);
  const parsed = parsePackages(nextPayload.packages);
  if (parsed.malformed) {
    return { payload: nextPayload, changes: [] };
  }

  const viewList = Array.isArray(parsed.packages.viewList) ? parsed.packages.viewList : [];
  const groups = Array.isArray(optimized.groups) ? optimized.groups : [];
  const changes: JsonRecord[] = [];

  for (const groupUpdate of groups) {
    if (!isRecord(groupUpdate)) continue;
    const groupIndex = groupUpdate.index;
    if (typeof groupIndex !== "number" || !Number.isInteger(groupIndex) || groupIndex < 0 || groupIndex >= viewList.length) {
      continue;
    }
    const targetGroup = viewList[groupIndex];
    if (!isRecord(targetGroup)) continue;

    const newGroupName = cleanString(groupUpdate.groupName);
    const oldGroupName = cleanString(targetGroup.groupName);
    if (newGroupName && newGroupName !== oldGroupName) {
      targetGroup.groupName = newGroupName;
      changes.push({
        path: `packages.viewList[${groupIndex}].groupName`,
        before: oldGroupName,
        after: newGroupName,
      });
    }

    const itemUpdates = Array.isArray(groupUpdate.items) ? groupUpdate.items : [];
    const targetItems = Array.isArray(targetGroup.list) ? targetGroup.list : [];
    for (const itemUpdate of itemUpdates) {
      if (!isRecord(itemUpdate)) continue;
      const itemIndex = itemUpdate.index;
      if (typeof itemIndex !== "number" || !Number.isInteger(itemIndex) || itemIndex < 0 || itemIndex >= targetItems.length) {
        continue;
      }
      const targetItem = targetItems[itemIndex];
      if (!isRecord(targetItem)) continue;
      const newTitle = cleanString(itemUpdate.title);
      const oldTitle = cleanString(targetItem.title);
      if (newTitle && newTitle !== oldTitle) {
        targetItem.title = newTitle;
        changes.push({
          path: `packages.viewList[${groupIndex}].list[${itemIndex}].title`,
          before: oldTitle,
          after: newTitle,
        });
      }
    }
  }

  nextPayload.packages = encodePackages(parsed.packages, parsed.wasString);
  return { payload: nextPayload, changes };
}

export function normalizeSupplyGoodsForLinKe(
  payload: JsonRecord,
  linKeMapping: JsonRecord,
  rbImageBaseUrl = "",
): JsonRecord {
  const groups = normalizeItemGroups(payload.packages);
  const merchantName = cleanString(payload.hostName)
    || entityText(payload.rbhost)
    || entityText(payload.company)
    || cleanString(payload.hostNameInput);
  const product: JsonRecord = {
    source: {
      type: "rebuild_supply_goods",
      id: cleanString(payload.SupplyGoodsId) || cleanString(payload.goodsId),
    },
    title: cleanString(payload.goodsName),
    salePrice: parseNumber(payload.price),
    originPrice: parseNumber(payload.originPrice),
    category: {
      id: cleanString(linKeMapping.categoryId),
      name: cleanString(linKeMapping.categoryName),
    },
    productType: parseIntValue(linKeMapping.productType),
    grouponType: cleanString(payload.majorType),
    images: [
      ...normalizeImages(payload.mainPic, rbImageBaseUrl),
      ...normalizeImages(payload.rbimages, rbImageBaseUrl),
    ],
    detailImages: normalizeImages(payload.detailImages, rbImageBaseUrl),
    description: cleanString(payload.details),
    features: cleanString(payload.goodsFeatures),
    stockQty: { totalStock: parseIntValue(payload.signAmount) },
    saleTime: {
      startDate: cleanString(payload.saleBegin),
      endDate: cleanString(payload.saleUntil),
    },
    validityPeriod: { endDate: cleanString(payload.validUntil) },
    itemGroups: groups,
    purchaseNotice: { additionalNotes: cleanString(payload.guideline) },
    merchant: { name: merchantName },
    hosts: merchantName ? [{ name: merchantName }] : [],
    fieldSources: {
      linKeProductType: cleanString(linKeMapping.mealTypeText),
      linKeCategory: cleanString(linKeMapping.classificationKey),
    },
    missingFields: [],
    warnings: [],
  };

  product.images = dedupeImages(product.images);
  product.detailImages = dedupeImages(product.detailImages);
  product.missingFields = collectMissingFields(product);
  return product;
}

export function normalizeItemGroups(packagesValue: unknown): JsonRecord[] {
  const { packages } = parsePackages(packagesValue);
  const viewList = Array.isArray(packages.viewList) ? packages.viewList : [];
  const groups: JsonRecord[] = [];
  for (const [groupIndex, group] of viewList.entries()) {
    if (!isRecord(group)) continue;
    const rawItems = Array.isArray(group.list) ? group.list : [];
    const items: JsonRecord[] = [];
    for (const [itemIndex, item] of rawItems.entries()) {
      if (!isRecord(item)) continue;
      items.push({
        id: cleanString(item.id) || `${groupIndex}-${itemIndex}`,
        name: cleanString(item.title),
        price: parseNumber(item.price),
        quantity: { amount: parseIntValue(item.num) || 1, unit: "FEN" },
      });
    }
    groups.push({
      id: cleanString(group.groupId) || String(groupIndex),
      name: cleanString(group.groupName),
      items,
      selectionRule: {
        totalCount: parseIntValue(group.groupSelectNum) || items.length,
        optionCount: items.length,
      },
      canRepeat: false,
    });
  }
  return groups;
}

export function normalizeImages(value: unknown, rbImageBaseUrl = ""): JsonRecord[] {
  if (value === null || value === undefined) return [];
  const rawItems = Array.isArray(value) ? value : [value];
  const images: JsonRecord[] = [];

  for (const [index, item] of rawItems.entries()) {
    let url = "";
    let uri = "";
    let name = "";
    if (typeof item === "string") {
      const raw = item.trim();
      if (raw.startsWith("http://") || raw.startsWith("https://")) {
        url = raw;
      } else if (rbImageBaseUrl) {
        url = new URL(raw.replace(/^\/+/, ""), rbImageBaseUrl.replace(/\/+$/, "") + "/").toString();
      } else {
        uri = raw;
      }
      name = raw ? raw.split("/").pop() ?? "" : "";
    } else if (isRecord(item)) {
      url = cleanString(item.url);
      uri = cleanString(item.uri);
      name = cleanString(item.name);
    }
    if (url || uri) {
      images.push({
        url,
        uri,
        name,
        sortableOnlyId: uri || url || String(index),
      });
    }
  }

  return images;
}

export function dedupeImages(images: unknown): JsonRecord[] {
  if (!Array.isArray(images)) return [];
  const seen = new Set<string>();
  const result: JsonRecord[] = [];
  for (const image of images) {
    if (!isRecord(image)) continue;
    const key = cleanString(image.uri) || cleanString(image.url);
    if (!key || seen.has(key)) continue;
    seen.add(key);
    result.push(image);
  }
  return result;
}

export function collectMissingFields(product: JsonRecord): string[] {
  const missing: string[] = [];
  if (!cleanString(product.title)) missing.push("title");
  if (!product.salePrice) missing.push("salePrice");
  if (!product.originPrice) missing.push("originPrice");
  if (!Array.isArray(product.images) || product.images.length === 0) missing.push("images");
  if (!Array.isArray(product.itemGroups) || product.itemGroups.length === 0) missing.push("itemGroups");
  const merchant = isRecord(product.merchant) ? product.merchant : {};
  const category = isRecord(product.category) ? product.category : {};
  if (!cleanString(merchant.name)) missing.push("merchant.name");
  if (!category.id) missing.push("category.id");
  if (!product.productType) missing.push("productType");
  return missing;
}
